File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.cpp
class Solution {
public:
    string removeSubstrings(string &s, const string& match)
    {
        // i for writing and j for reading
        int i = 0, j = 0;
        while (j < s.length()) {
            s[i] = s[j];
            i++;

            if(i >= 2 && s[i - 2] == match[0] && s[i - 1] == match[1]) {
                i -= 2;
            }

            j++;
        }
        return s.substr(0, i);
    }

    int maximumGain(string s, int x, int y) {
        string maxString = (x >= y) ? "ab" : "ba";
        string minString = (x < y) ? "ab" : "ba";
        int maxPoint = max(x, y);
        int minPoint = min(x, y);

        string afterFirstPass = removeSubstrings(s, maxString);
        int firstGain = ((int)s.size() - (int)afterFirstPass.size()) / 2 * maxPoint;

        string afterSecondPass = removeSubstrings(afterFirstPass, minString);
        int secondGain = ((int)afterFirstPass.size() - (int)afterSecondPass.size()) / 2 * minPoint;

        return firstGain + secondGain;
    }
};

File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.java
class Solution {

    private String removeSubstrings(String s, String match)
    {
        StringBuilder sb = new StringBuilder(s);
        // i for writing and j for reading
        int i = 0, j = 0;
        while (j < s.length()) {
            sb.setCharAt(i, s.charAt(j));
            i++;

            if(i >= 2 && sb.charAt(i - 2) == match.charAt(0) && sb.charAt(i - 1) == match.charAt(1)) {
                i -= 2;
            }

            j++;
        }
        return sb.substring(0, i);
    }

    public int maximumGain(String s, int x, int y)
    {
        String maxString = x >= y ? "ab" : "ba";
        String minString = x < y ? "ab" : "ba";

        int maxPoint = Math.max(x, y);
        int minPoint = Math.min(x, y);

        int score = 0;
        String stringAfterOnePass = removeSubstrings(s, maxString);
        score += ((s.length() - stringAfterOnePass.length()) / 2) * maxPoint;

        String stringAfterTwoPass = removeSubstrings(stringAfterOnePass, minString);
        score += ((stringAfterOnePass.length() - stringAfterTwoPass.length()) / 2) * minPoint;

        return score;
    }

}